### PR TITLE
Closes AgileVentures/MetPlus_tracker#438

### PR DESCRIPTION
**Pull request by nchristian777**
-Adds authentication to job applications controller.
-Adds job_application_policy.rb and corresponding test job_application_policy_spec.rb
-Adds authentication tests to job_applications_controller_spec.rb
-Refactors job_applications_controller.rb and job_applications_controller_spec.rb to comply with Rubocop standards.

### CODE REVIEW

File: app/policies/job_application_policy.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+  end
+
+  def show?
+    company_person? user
+  end
+
+  private
+
+  def company_person?(user)
+    User.is_company_person? user
```
Comment (joaopapereira) [inline app/policies/job_application_policy.rb]: I believe this validation is not enough. The Company of the user need to be the same as the company the job application is to.

Let's imagine the Job Seeker Apply to Company 1 and Company 2.
Company 2 people cannot see that the Job Seeker applied to Company 1
Company 1 people cannot see that the Job Seeker applied to Company 2

Comment (joaopapereira) [inline app/policies/job_application_policy.rb]: It should be something similar to
def company_person?(user)
  User.is_company_person? user and user.pets_user.company == resource.job.company
end

File: spec/policies/job_application_policy_spec.rb [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  let(:company_admin) { FactoryGirl.create(:company_admin, company: company) }
+  let(:company_contact) do
+    FactoryGirl.create(:company_contact,
+                       company: company)
+  end
+  let(:job_seeker) { FactoryGirl.create(:job_seeker) }
+
+  permissions :accept?, :reject?, :show? do
+    it 'denies access if user is not a company admin/contact' do
+      expect(JobApplicationPolicy).not_to permit(job_seeker)
```
Comment (joaopapereira) [inline spec/policies/job_application_policy_spec.rb]: We need to add tests with the following users:
* Agency Admin
* Job Developer
* Case Manager
* Company Admin from another company
* Company Contact from another company

File: spec/controllers/job_applications_controller_spec.rb [hunk 3]
```diff
@@ -1,69 +1,153 @@
...
+        context 'company_admin' do
+          let(:request) { patch :accept, id: company_admin }
+          it_behaves_like 'unauthenticated request'
+        end
+        context 'company_contact' do
+          let(:request) { patch :accept, id: company_contact }
+          it_behaves_like 'unauthenticated request'
+        end
+      end
+      context 'Job Seeker' do
```
Comment (joaopapereira) [inline spec/controllers/job_applications_controller_spec.rb]: We need to duplicate this context with the following users:
* Agency Admin
* Job Developer
* Case Manager
* Company Admin from another company
* Company Contact from another company

Comment (joaopapereira) [inline spec/controllers/job_applications_controller_spec.rb]: We also need to check that all the actions are tests, so we are missing

- PATCH :reject
- GET Show

File: app/controllers/job_applications_controller.rb [hunk 4]
```diff
@@ -1,54 +1,64 @@
...
-			redirect_back_or_default
-		end
-	end
+  def reject_success_response(request)
+    if request.xhr?
+      render json: { message: 'Job application rejected', status: 200 }
+    else
+      flash[:notice] = 'Job application rejected.'
+      redirect_to controller: 'jobs', action: 'applications',
+                  id: @job_application.job.id unless request.xhr?
```
Comment (patmbolger) [inline app/controllers/job_applications_controller.rb]: Not part of this story but might as well clean this up.  That is, drop this from the statement:
```
unless request.xhr?
```

**Comment by nchristian777** [comment]
Made the following improvements based on feedback

- job_application_policy denies company people who are from a different company
- added tests in job_application_policy_spec.rb which include the following
- - tests for agency admin, job developer, case manager
- - tests for company people form the wrong company
- added tests in job_application_controller_spec which include the following
- - tests for agency admin, job developer, case manager
- - tests for company people form the wrong company
- - tests for PATCH reject and GET show

**Comment by patmbolger** [comment]
Changes look good to me - note that I did not check that all of João's concerns were addressed.

**Comment by joaopapereira** [comment]
Going to merge

**Pull request merged**